File: obel/obeltests/set.m
//
//  set.m
//  obeltests
//
//  Created by Jack Mousseau on 11/29/21.
//

#import <obel/set.h>
#import <XCTest/XCTest.h>

@interface ObelSetTests : XCTestCase

@end

@implementation ObelSetTests

- (void)testSparseInsert {
    uint16_t sparse[4] = { 0, 1, 0, 0 };
    uint16_t dense[4] = { 3, 1, 0, 0};

    obel_sparse_set_t set = {
        .capacity = 4,
        .count = 2,
        .sparse = sparse,
        .dense = dense
    };

    obel_sparse_set_insert(&set, 0);
    XCTAssertEqual(set.count, 3);
    XCTAssertEqual(set.dense[2], 0);
    XCTAssertEqual(set.sparse[0], 2);

    obel_sparse_set_insert(&set, 1);
    XCTAssertEqual(set.count, 3);
}

- (void)testSparseRemove {
    uint16_t sparse[4] = { 0, 1, 0, 0 };
    uint16_t dense[4] = { 3, 1, 0, 0};

    obel_sparse_set_t set = {
        .capacity = 4,
        .count = 2,
        .sparse = sparse,
        .dense = dense
    };

    obel_sparse_set_remove(&set, 3);
    XCTAssertEqual(set.count, 1);
    XCTAssertEqual(set.dense[0], 1);
    XCTAssertEqual(set.sparse[1], 0);

    obel_sparse_set_remove(&set, 3);
    XCTAssertEqual(set.count, 1);

    obel_sparse_set_remove(&set, 2);
    XCTAssertEqual(set.count, 1);
}

- (void)testSparseContains {
    uint16_t sparse[4] = { 0, 1, 0, 0 };
    uint16_t dense[4] = { 3, 1, 0, 0};

    obel_sparse_set_t set = {
        .capacity = 4,
        .count = 2,
        .sparse = sparse,
        .dense = dense
    };

    XCTAssertFalse(obel_sparse_set_contains(&set, 0));
    XCTAssertTrue(obel_sparse_set_contains(&set, 1));
    XCTAssertFalse(obel_sparse_set_contains(&set, 2));
    XCTAssertTrue(obel_sparse_set_contains(&set, 3));
}

- (void)testSparseSetClear {
    uint16_t sparse[4] = { 0, 1, 0, 0 };
    uint16_t dense[4] = { 3, 1, 0, 0};

    obel_sparse_set_t set = {
        .capacity = 4,
        .count = 2,
        .sparse = sparse,
        .dense = dense
    };

    obel_sparse_set_clear(&set);
    XCTAssertEqual(set.count, 0);
    XCTAssertFalse(obel_sparse_set_contains(&set, 0));
    XCTAssertFalse(obel_sparse_set_contains(&set, 1));
    XCTAssertFalse(obel_sparse_set_contains(&set, 2));
    XCTAssertFalse(obel_sparse_set_contains(&set, 3));
}

@end

File: obel/set.c
//
//  set.c
//  obel
//
//  Created by Jack Mousseau on 11/29/21.
//

#include <obel/set.h>

OBEL_SPARSE_FUNC void obel_sparse_set_insert(obel_sparse_set_t *set, uint16_t element) {
    OBEL_ASSERT(element < set->capacity);
    if (!obel_sparse_set_contains(set, element)) {
        uint16_t dense = set->count++;
        set->dense[dense] = element;
        set->sparse[element] = dense;
    }
}

OBEL_SPARSE_FUNC void obel_sparse_set_remove(obel_sparse_set_t *set, uint16_t element) {
    OBEL_ASSERT(element < set->capacity);
    if (obel_sparse_set_contains(set, element)) {
        uint16_t dense = set->sparse[element];
        uint16_t last = set->dense[--set->count];
        set->dense[dense] = last;
        set->sparse[last] = dense;
    }
}

OBEL_SPARSE_FUNC bool obel_sparse_set_contains(obel_sparse_set_t *set, uint16_t element) {
    OBEL_ASSERT(element < set->capacity);
    uint16_t dense = set->sparse[element];
    return (dense < set->count && set->dense[dense] == element);
}

OBEL_SPARSE_FUNC void obel_sparse_set_clear(obel_sparse_set_t *set) {
    set->count = 0;
}
